sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -55,100 +55,81 @@ object ConstantFolding extends Rule[LogicalPlan] {
-        (None, Seq(((right, left), e)))
-      case a: And =>
-        val (newLeft, equalityPredicatesLeft) = traverse(a.left, replaceChildren = false)
-        val (newRight, equalityPredicatesRight) = traverse(a.right, replaceChildren = false)
-        val equalityPredicates = equalityPredicatesLeft ++ equalityPredicatesRight
-        val newSelf = if (equalityPredicates.nonEmpty && replaceChildren) {
-          Some(And(replaceConstants(newLeft.getOrElse(a.left), equalityPredicates),
-            replaceConstants(newRight.getOrElse(a.right), equalityPredicates)))
-        } else {
-          if (newLeft.isDefined || newRight.isDefined) {
-            Some(And(newLeft.getOrElse(a.left), newRight.getOrElse(a.right)))
-          } else {
-            None
-          }
-        }
-        (newSelf, equalityPredicates)
-      case o: Or =>
-        // Ignore the EqualityPredicates from children since they are only propagated through And.
-        val (newLeft, _) = traverse(o.left, replaceChildren = true)
-        val (newRight, _) = traverse(o.right, replaceChildren = true)
-        val newSelf = if (newLeft.isDefined || newRight.isDefined) {
-          Some(Or(left = newLeft.getOrElse(o.left), right = newRight.getOrElse((o.right))))
+  private def traverse(expression: Expression): (Expression, Map[Expression, Literal]) =
+    expression match {
+      case e @ EqualTo(left, right: Literal) if !left.foldable && left.deterministic =>
+        (e, Map(left.canonicalized -> right))
+      case e @ EqualTo(left: Literal, right) if !right.foldable && right.deterministic =>
+        (e, Map(right.canonicalized -> left))
+      case e @ EqualNullSafe(left, right: Literal) if !left.foldable && left.deterministic =>
+        (e, Map(left.canonicalized -> right))
+      case e @ EqualNullSafe(left: Literal, right) if !right.foldable && right.deterministic =>
+        (e, Map(right.canonicalized -> left))
+      case a @ And(left, right) =>
+        val (newLeft, equalityPredicatesLeft) = traverse(left)
+        val replacedRight = replaceConstants(right, equalityPredicatesLeft)
+        val (replacedNewRight, equalityPredicatesRight) = traverse(replacedRight)
+        val replacedNewLeft = replaceConstants(newLeft, equalityPredicatesRight)
+        val newAnd = if ((replacedNewLeft fastEquals left) && (replacedNewRight fastEquals right)) {
+          a
         } else {
-          None
+          And(replacedNewLeft, replacedNewRight)
         }
-        (newSelf, Seq.empty)
-      case n: Not =>
-        // Ignore the EqualityPredicates from children since they are only propagated through And.
-        val (newChild, _) = traverse(n.child, replaceChildren = true)
-        (newChild.map(Not), Seq.empty)
-      case _ => (None, Seq.empty)
+        (newAnd, equalityPredicatesLeft ++= equalityPredicatesRight)
+      case o @ (_: Or | _: Not) =>
+        (o.mapChildren(traverse(_)._1), Map.empty)
+      case _ => (expression, Map.empty)
```

Comment:
The original version stops traversing if it encounters a node other than `And`, `Or` or `Not`, but I don't see the reason why. Any thoughts?

Reply:
Actually, I don't see why do we stop traversing at all. For example `b` in `If(a = b AND b = 1, ..., ...)` doesn't get "resolved" now.

Reply:
well, `If` can be present also in other nodes than `Filter` and I think it should be handled separately from here, as we cannot mix what's inside the in clause an what's a filter...

Reply:
This rule always start traversing from a condition expression of a `Filter` and never replace anything that is outside of that condition expression. I still don't see why we stop propagation as soon as we leave the child subtree of `And`, `Or` and `Not` nodes of the condition. I think there can be other subtrees of `And` nodes deep under the filter condition that could leverage constant propagation among them like in my example above.

Reply:
On a second thought, a follow up PR with more examples could address this.
Thanks @mgaido91.